File: jetbrains/src/main/kotlin/dev/pacmon/jetbrains/core/MixManifest.kt
package dev.pacmon.jetbrains.core

private enum class MixTokenKind { WORD, ATOM, LITERAL, SYMBOL }

private data class MixToken(
    val kind: MixTokenKind,
    val value: String,
    val start: Int,
    val end: Int,
    val contentStart: Int? = null,
    val contentEnd: Int? = null,
)

object MixManifestAdapter : ManifestAdapter {
    override val kind = ManifestKind.MIX
    override val fileNames = listOf("mix.exs")
    override val notesRelativePath = ".pacmon/mix/DEPENDENCY-NOTES.md"
    override fun normalizeNoteKey(raw: String): String = ManifestRegistry.stripName(raw)

    private val openToClose = mapOf('(' to ')', '[' to ']', '{' to '}')

    override fun extractDependencies(text: String): List<DependencyEntry> {
        val tokens = tokenize(text)
        val out = mutableListOf<DependencyEntry>()
        val seen = mutableSetOf<String>()
        for (open in literalListStarts(tokens)) {
            val close = matchingToken(tokens, open) ?: continue
            for (dependency in dependenciesFromList(tokens, open, close)) {
                if (seen.add("${dependency.scope}\u0000${dependency.noteKey}")) out.add(dependency)
            }
        }
        return out
    }

    private fun isWordStart(char: Char?): Boolean = char != null && (char == '_' || char.isLetter())
    private fun isWordPart(char: Char?): Boolean = char != null && (char == '_' || char == '@' || char == '!' || char == '?' || char.isLetterOrDigit())

    private fun decodeQuoted(value: String): String = buildString {
        var cursor = 0
        while (cursor < value.length) {
            val char = value[cursor++]
            if (char != '\\' || cursor >= value.length) {
                append(char)
                continue
            }
            when (val escaped = value[cursor++]) {
                'n' -> append('\n')
                'r' -> append('\r')
                't' -> append('\t')
                else -> append(escaped)
            }
        }
    }

    private fun quotedEnd(text: String, start: Int, quote: Char): Int {
        val marker = if (text.startsWith(quote.toString().repeat(3), start)) quote.toString().repeat(3) else quote.toString()
        var cursor = start + marker.length
        while (cursor < text.length) {
            if (text.startsWith(marker, cursor)) return cursor + marker.length
            cursor += if (text[cursor] == '\\' && cursor + 1 < text.length) 2 else 1
        }
        return text.length
    }

    private fun sigilEnd(text: String, start: Int): Int? {
        if (text.getOrNull(start) != '~' || text.getOrNull(start + 1)?.isLetter() != true) return null
        val delimiterStart = start + 2
        val opening = text.getOrNull(delimiterStart) ?: return null
        if (opening.isLetterOrDigit() || opening == '_' || opening.isWhitespace()) return null
        if ((opening == '"' || opening == '\'') && text.startsWith(opening.toString().repeat(3), delimiterStart)) {
            var cursor = quotedEnd(text, delimiterStart, opening)
            while (text.getOrNull(cursor)?.isLetter() == true) cursor++
            return cursor
        }
        val closing = mapOf('(' to ')', '[' to ']', '{' to '}', '<' to '>')[opening] ?: opening
        var depth = if (opening == closing) 0 else 1
        var cursor = delimiterStart + 1
        while (cursor < text.length) {
            val char = text[cursor]
            if (char == '\\' && cursor + 1 < text.length) {
                cursor += 2
                continue
            }
            if (opening != closing && char == opening) depth++
            if (char == closing) {
                depth--
                if (opening == closing || depth == 0) {
                    cursor++
                    while (text.getOrNull(cursor)?.isLetter() == true) cursor++
                    return cursor
                }
            }
            cursor++
        }
        return text.length
    }

    private fun tokenize(text: String): List<MixToken> {
        val tokens = mutableListOf<MixToken>()
        var cursor = 0
        while (cursor < text.length) {
            val char = text[cursor]
            if (char.isWhitespace()) {
                cursor++
                continue
            }
            if (char == '#') {
                while (cursor < text.length && text[cursor] != '\n') cursor++
                continue
            }
            val endOfSigil = sigilEnd(text, cursor)
            if (endOfSigil != null) {
                tokens.add(MixToken(MixTokenKind.LITERAL, text.substring(cursor, endOfSigil), cursor, endOfSigil))
                cursor = endOfSigil
                continue
            }
            if (char == ':' && text.getOrNull(cursor + 1) in setOf('"', '\'')) {
                val quote = text[cursor + 1]
                val quotedStart = cursor + 1
                val end = quotedEnd(text, quotedStart, quote)
                val markerWidth = if (text.startsWith(quote.toString().repeat(3), quotedStart)) 3 else 1
                val contentStart = quotedStart + markerWidth
                val contentEnd = (end - markerWidth).coerceAtLeast(contentStart)
                tokens.add(
                    MixToken(
                        MixTokenKind.ATOM,
                        decodeQuoted(text.substring(contentStart, contentEnd)),
                        cursor,
                        end,
                        contentStart,
                        contentEnd,
                    ),
                )
                cursor = end
                continue
            }
            if (char == ':' && isWordStart(text.getOrNull(cursor + 1))) {
                var end = cursor + 2
                while (isWordPart(text.getOrNull(end))) end++
                tokens.add(MixToken(MixTokenKind.ATOM, text.substring(cursor + 1, end), cursor, end, cursor + 1, end))
                cursor = end
                continue
            }
            if (char == '"' || char == '\'') {
                val end = quotedEnd(text, cursor, char)
                tokens.add(MixToken(MixTokenKind.LITERAL, text.substring(cursor, end), cursor, end))
                cursor = end
                continue
            }
            if (isWordStart(char)) {
                var end = cursor + 1
                while (isWordPart(text.getOrNull(end))) end++
                tokens.add(MixToken(MixTokenKind.WORD, text.substring(cursor, end), cursor, end))
                cursor = end
                continue
            }
            tokens.add(MixToken(MixTokenKind.SYMBOL, char.toString(), cursor, cursor + 1))
            cursor++
        }
        return tokens
    }

    private fun matchingToken(tokens: List<MixToken>, openIndex: Int): Int? {
        val wanted = tokens.getOrNull(openIndex)?.value?.singleOrNull()?.let(openToClose::get) ?: return null
        val stack = mutableListOf(wanted)
        for (index in openIndex + 1 until tokens.size) {
            val value = tokens[index].value.singleOrNull()
            val nested = value?.let(openToClose::get)
            if (nested != null) stack.add(nested)
            else if (value == stack.lastOrNull()) {
                stack.removeAt(stack.lastIndex)
                if (stack.isEmpty()) return index
            }
        }
        return null
    }

    private fun literalListStarts(tokens: List<MixToken>): List<Int> {
        val starts = sortedSetOf<Int>()
        val ancestors = mutableListOf<Char>()
        for (index in tokens.indices) {
            val token = tokens[index]
            if (
                '}' !in ancestors &&
                token.kind == MixTokenKind.WORD && token.value == "deps" &&
                tokens.getOrNull(index + 1)?.value == ":" && tokens.getOrNull(index + 2)?.value == "["
            ) {
                starts.add(index + 2)
            }
            if (token.kind == MixTokenKind.WORD && token.value in setOf("def", "defp")) {
                if (tokens.getOrNull(index + 1)?.kind == MixTokenKind.WORD && tokens[index + 1].value == "deps") {
                    var cursor = index + 2
                    if (tokens.getOrNull(cursor)?.value == "(") {
                        val close = matchingToken(tokens, cursor)
                        cursor = if (close != null && close == cursor + 1) close + 1 else -1
                    }
                    if (tokens.getOrNull(cursor)?.value == ",") cursor++
                    if (tokens.getOrNull(cursor)?.value == "do") {
                        cursor++
                        if (tokens.getOrNull(cursor)?.value == ":") cursor++
                        if (tokens.getOrNull(cursor)?.value == "[") starts.add(cursor)
                    }
                }
            }
            val value = token.value.singleOrNull()
            val nested = value?.let(openToClose::get)
            if (nested != null) ancestors.add(nested)
            else if (value != null && value == ancestors.lastOrNull()) ancestors.removeAt(ancestors.lastIndex)
        }
        return starts.toList()
    }

    private fun atomsFromOption(tokens: List<MixToken>, name: String): List<String>? {
        val stack = mutableListOf<Char>()
        for (index in 0 until (tokens.size - 2).coerceAtLeast(0)) {
            val token = tokens[index]
            if (stack.isEmpty() && token.kind == MixTokenKind.WORD && token.value == name && tokens[index + 1].value == ":") {
                val value = tokens[index + 2]
                if (value.kind == MixTokenKind.ATOM) return listOf(value.value)
                if (value.value != "[") return null
                val close = matchingToken(tokens, index + 2) ?: return null
                val atoms = mutableListOf<String>()
                for (itemIndex in index + 3 until close) {
                    val item = tokens[itemIndex]
                    if (item.kind == MixTokenKind.ATOM) atoms.add(item.value)
                    else if (item.value != ",") return null
                }
                return if (atoms.isEmpty()) null else atoms
            }
            val symbol = token.value.singleOrNull()
            val nested = symbol?.let(openToClose::get)
            if (nested != null) stack.add(nested)
            else if (symbol != null && symbol == stack.lastOrNull()) stack.removeAt(stack.lastIndex)
        }
        return null
    }

    private fun dependencyFromTuple(tokens: List<MixToken>): DependencyEntry? {
        if (tokens.size < 4 || tokens.first().value != "{" || tokens.last().value != "}") return null
        val atom = tokens[1]
        if (atom.kind != MixTokenKind.ATOM || tokens[2].value != ",") return null
        val options = tokens.subList(3, tokens.lastIndex)
        val environments = atomsFromOption(options, "only")
        val targets = atomsFromOption(options, "targets")
        val scope = buildString {
            append("deps")
            if (environments != null) append(':').append(environments.joinToString(","))
            if (targets != null) append('@').append(targets.joinToString(","))
        }
        val primaryStart = atom.contentStart ?: atom.start
        val primaryEnd = atom.contentEnd ?: atom.end
        return dependencyEntry(
            atom.value,
            scope,
            SourceRange(primaryStart, primaryEnd - primaryStart),
            listOf(SourceRange(atom.start, atom.end - atom.start)),
            atom.value,
            SourceRange(tokens.first().start, 1),
        )
    }

    private fun dependenciesFromList(tokens: List<MixToken>, open: Int, close: Int): List<DependencyEntry> {
        val out = mutableListOf<DependencyEntry>()
        val stack = mutableListOf<Char>()
        var elementStart = open + 1
        fun visit(end: Int) {
            dependencyFromTuple(tokens.subList(elementStart, end))?.let(out::add)
        }
        for (index in open + 1 until close) {
            val value = tokens[index].value.singleOrNull()
            val nested = value?.let(openToClose::get)
            if (nested != null) stack.add(nested)
            else if (value == stack.lastOrNull()) stack.removeAt(stack.lastIndex)
            else if (value == ',' && stack.isEmpty()) {
                visit(index)
                elementStart = index + 1
            }
        }
        visit(close)
        return out
    }
}

File: src/core/mixManifest.ts
import { dependencyEntry } from './dependency';
import type { DependencyEntry, SourceRange } from './model';

type TokenKind = 'word' | 'atom' | 'literal' | 'symbol';

interface Token {
  kind: TokenKind;
  value: string;
  start: number;
  end: number;
  contentStart?: number;
  contentEnd?: number;
}

const OPEN_TO_CLOSE: Readonly<Record<string, string>> = {
  '(': ')',
  '[': ']',
  '{': '}',
};

function decodeQuoted(value: string, quote: string): string {
  if (quote === '"') {
    try {
      return JSON.parse(`"${value}"`) as string;
    } catch {
      // Keep malformed-but-readable names useful while the user is editing.
    }
  }
  return value.replace(/\\([\\'"nrt])/g, (_match, escaped: string) => {
    if (escaped === 'n') return '\n';
    if (escaped === 'r') return '\r';
    if (escaped === 't') return '\t';
    return escaped;
  });
}

function quotedEnd(text: string, start: number, quote: '"' | "'"): number {
  const triple = text.slice(start, start + 3) === quote.repeat(3);
  const marker = triple ? quote.repeat(3) : quote;
  let cursor = start + marker.length;
  while (cursor < text.length) {
    if (text.startsWith(marker, cursor)) return cursor + marker.length;
    if (text[cursor] === '\\' && cursor + 1 < text.length) cursor += 2;
    else cursor++;
  }
  return text.length;
}

function sigilEnd(text: string, start: number): number | undefined {
  if (text[start] !== '~' || !/[A-Za-z]/.test(text[start + 1] ?? '')) return undefined;
  const delimiterStart = start + 2;
  const opening = text[delimiterStart];
  if (!opening || /[A-Za-z0-9_\s]/.test(opening)) return undefined;
  if ((opening === '"' || opening === "'") && text.slice(delimiterStart, delimiterStart + 3) === opening.repeat(3)) {
    let cursor = quotedEnd(text, delimiterStart, opening);
    while (/[A-Za-z]/.test(text[cursor] ?? '')) cursor++;
    return cursor;
  }
  const closing = ({ '(': ')', '[': ']', '{': '}', '<': '>' } as Record<string, string>)[opening] ?? opening;
  let depth = opening === closing ? 0 : 1;
  let cursor = delimiterStart + 1;
  while (cursor < text.length) {
    const char = text[cursor]!;
    if (char === '\\' && cursor + 1 < text.length) {
      cursor += 2;
      continue;
    }
    if (opening !== closing && char === opening) depth++;
    if (char === closing) {
      if (opening === closing || --depth === 0) {
        cursor++;
        while (/[A-Za-z]/.test(text[cursor] ?? '')) cursor++;
        return cursor;
      }
    }
    cursor++;
  }
  return text.length;
}

/** A small Elixir lexer used only for statically visible Mix dependency lists. */
function tokenize(text: string): Token[] {
  const tokens: Token[] = [];
  let cursor = 0;
  while (cursor < text.length) {
    const char = text[cursor]!;
    if (/\s/.test(char)) {
      cursor++;
      continue;
    }
    if (char === '#') {
      while (cursor < text.length && text[cursor] !== '\n') cursor++;
      continue;
    }
    const endOfSigil = sigilEnd(text, cursor);
    if (endOfSigil !== undefined) {
      tokens.push({ kind: 'literal', value: text.slice(cursor, endOfSigil), start: cursor, end: endOfSigil });
      cursor = endOfSigil;
      continue;
    }
    if (char === ':' && (text[cursor + 1] === '"' || text[cursor + 1] === "'")) {
      const quote = text[cursor + 1] as '"' | "'";
      const quotedStart = cursor + 1;
      const end = quotedEnd(text, quotedStart, quote);
      const markerWidth = text.slice(quotedStart, quotedStart + 3) === quote.repeat(3) ? 3 : 1;
      const contentStart = quotedStart + markerWidth;
      const contentEnd = Math.max(contentStart, end - markerWidth);
      tokens.push({
        kind: 'atom',
        value: decodeQuoted(text.slice(contentStart, contentEnd), quote),
        start: cursor,
        end,
        contentStart,
        contentEnd,
      });
      cursor = end;
      continue;
    }
    if (char === ':') {
      const match = /^[A-Za-z_][A-Za-z0-9_@!?]*/.exec(text.slice(cursor + 1));
      if (match) {
        const end = cursor + 1 + match[0].length;
        tokens.push({
          kind: 'atom',
          value: match[0],
          start: cursor,
          end,
          contentStart: cursor + 1,
          contentEnd: end,
        });
        cursor = end;
        continue;
      }
    }
    if (char === '"' || char === "'") {
      const end = quotedEnd(text, cursor, char);
      tokens.push({ kind: 'literal', value: text.slice(cursor, end), start: cursor, end });
      cursor = end;
      continue;
    }
    const word = /^[A-Za-z_][A-Za-z0-9_@!?]*/.exec(text.slice(cursor));
    if (word) {
      const end = cursor + word[0].length;
      tokens.push({ kind: 'word', value: word[0], start: cursor, end });
      cursor = end;
      continue;
    }
    tokens.push({ kind: 'symbol', value: char, start: cursor, end: cursor + 1 });
    cursor++;
  }
  return tokens;
}

function matchingToken(tokens: readonly Token[], openIndex: number): number | undefined {
  const wanted = OPEN_TO_CLOSE[tokens[openIndex]?.value ?? ''];
  if (!wanted) return undefined;
  const stack = [wanted];
  for (let i = openIndex + 1; i < tokens.length; i++) {
    const value = tokens[i]!.value;
    const nested = OPEN_TO_CLOSE[value];
    if (nested) stack.push(nested);
    else if (value === stack[stack.length - 1]) {
      stack.pop();
      if (stack.length === 0) return i;
    }
  }
  return undefined;
}

function literalListStarts(tokens: readonly Token[]): number[] {
  const starts = new Set<number>();
  const ancestors: string[] = [];
  for (let i = 0; i < tokens.length; i++) {
    const token = tokens[i]!;
    if (
      !ancestors.includes('}')
      && token.kind === 'word'
      && token.value === 'deps'
      && tokens[i + 1]?.value === ':'
      && tokens[i + 2]?.value === '['
    ) {
      starts.add(i + 2);
    }
    if (token.kind === 'word' && (token.value === 'def' || token.value === 'defp')) {
      if (tokens[i + 1]?.kind === 'word' && tokens[i + 1]?.value === 'deps') {
        let cursor = i + 2;
        if (tokens[cursor]?.value === '(') {
          const close = matchingToken(tokens, cursor);
          if (close !== undefined && close === cursor + 1) cursor = close + 1;
          else cursor = -1;
        }
        if (tokens[cursor]?.value === ',') cursor++;
        if (tokens[cursor]?.value === 'do') {
          cursor++;
          if (tokens[cursor]?.value === ':') cursor++;
          if (tokens[cursor]?.value === '[') starts.add(cursor);
        }
      }
    }
    const nested = OPEN_TO_CLOSE[token.value];
    if (nested) ancestors.push(nested);
    else if (token.value === ancestors[ancestors.length - 1]) ancestors.pop();
  }
  return [...starts].sort((a, b) => a - b);
}

function atomsFromOption(tokens: readonly Token[], name: string): string[] | undefined {
  const stack: string[] = [];
  for (let i = 0; i < tokens.length - 2; i++) {
    const token = tokens[i]!;
    if (stack.length === 0 && token.kind === 'word' && token.value === name && tokens[i + 1]?.value === ':') {
      const value = tokens[i + 2]!;
      if (value.kind === 'atom') return [value.value];
      if (value.value !== '[') return undefined;
      const close = matchingToken(tokens, i + 2);
      if (close === undefined) return undefined;
      const atoms: string[] = [];
      for (let j = i + 3; j < close; j++) {
        const item = tokens[j]!;
        if (item.kind === 'atom') atoms.push(item.value);
        else if (item.value !== ',') return undefined;
      }
      return atoms.length > 0 ? atoms : undefined;
    }
    const nested = OPEN_TO_CLOSE[token.value];
    if (nested) stack.push(nested);
    else if (token.value === stack[stack.length - 1]) stack.pop();
  }
  return undefined;
}

function dependencyFromTuple(tokens: readonly Token[]): DependencyEntry | undefined {
  if (tokens.length < 4 || tokens[0]?.value !== '{' || tokens[tokens.length - 1]?.value !== '}') return undefined;
  const atom = tokens[1];
  if (!atom || atom.kind !== 'atom' || tokens[2]?.value !== ',') return undefined;
  const options = tokens.slice(3, -1);
  const environments = atomsFromOption(options, 'only');
  const targets = atomsFromOption(options, 'targets');
  let scope = 'deps';
  if (environments) scope += `:${environments.join(',')}`;
  if (targets) scope += `@${targets.join(',')}`;
  const primaryRange: SourceRange = {
    offset: atom.contentStart ?? atom.start,
    length: (atom.contentEnd ?? atom.end) - (atom.contentStart ?? atom.start),
  };
  return dependencyEntry(
    atom.value,
    scope,
    primaryRange,
    [{ offset: atom.start, length: atom.end - atom.start }],
    atom.value,
    { offset: tokens[0]!.start, length: 1 },
  );
}

function dependenciesFromList(tokens: readonly Token[], open: number, close: number): DependencyEntry[] {
  const out: DependencyEntry[] = [];
  let elementStart = open + 1;
  const stack: string[] = [];
  const visit = (end: number): void => {
    const element = tokens.slice(elementStart, end);
    const dependency = dependencyFromTuple(element);
    if (dependency) out.push(dependency);
  };
  for (let i = open + 1; i < close; i++) {
    const value = tokens[i]!.value;
    const nested = OPEN_TO_CLOSE[value];
    if (nested) stack.push(nested);
    else if (value === stack[stack.length - 1]) stack.pop();
    else if (value === ',' && stack.length === 0) {
      visit(i);
      elementStart = i + 1;
    }
  }
  visit(close);
  return out;
}

/** Extracts direct, statically visible Mix dependencies without evaluating mix.exs. */
export function extractMixDependencies(text: string): DependencyEntry[] {
  const tokens = tokenize(text);
  const out: DependencyEntry[] = [];
  const seen = new Set<string>();
  for (const open of literalListStarts(tokens)) {
    const close = matchingToken(tokens, open);
    if (close === undefined) continue;
    for (const dependency of dependenciesFromList(tokens, open, close)) {
      const key = `${dependency.scope}\0${dependency.noteKey}`;
      if (seen.has(key)) continue;
      seen.add(key);
      out.push(dependency);
    }
  }
  return out;
}
